File: sap/sap/report/product_order_script_report/product_order_script_report.py
# Copyright (c) 2023, ahmed and contributors
# For license information, please see license.txt

import frappe


def execute(filters=None):
	columns, data = [], []
	columns = get_columns(filters)
	data = get_data(filters)
	
	return columns, data


def get_columns(filters=None):
	# print(filters)
	col= [
				{
					'fieldname': 'product_order',
					'label': ('Product Order'),
					'fieldtype': 'Link',
					'options': 'Product Order',
					'width':'200'
				},
				{
				    'fieldname': 'item_serial',
				    'label': ('Item Serial'),
				    'fieldtype': 'Data',
				    # 'options': 'Product Order Details'
					'width':'200'
				},
				{
				    'fieldname': 'pallet_no',
				    'label': ('Pallet No'),
				    'fieldtype': 'Data',
				    # 'options': 'Product Order Details'
					'width':'130'
				},
				{
				    'fieldname': 'quality_status',
				    'label': ('Quality Status'),
				    'fieldtype': 'Data',
				    # 'options': 'Product Order Details'
					'width':'200'
				},
				{
				    'fieldname': 'item_status',
				    'label': ('	Item Status	'),
				    'fieldtype': 'Data',
				    # 'options': 'Product Order Details'
					'width':'150'
				},
				{
				    'fieldname': 'creation_date',
				    'label': ('Creation Date'),
				    'fieldtype': 'Date',
				    # 'options': 'Product Order Details'
					'width':'200'
				},
				{
				    'fieldname': 'sap_serial',
				    'label': ('Sap Serial Number'),
				    'fieldtype': 'Data',
				    # 'options': 'Product Order Details'
					'width':'150'
				},
				{
				    'fieldname': 'qt_inspection_lab',
				    'label': ('Qt Inspection lab'),
				    'fieldtype': 'Data',
				    # 'options': 'Product Order Details'
					'width':'150'
				},
				{
				    'fieldname': 'qt_inspection_yard',
				    'label': ('Qt Inspection yard'),
				    'fieldtype': 'Data',
				    # 'options': 'Product Order Details'
					'width':'150'
				},
				{
				    'fieldname': 'quality_status_lab',
				    'label': ('Quality Status lab'),
				    'fieldtype': 'Data',
				    # 'options': 'Product Order Details'
					'width':'150'
				},
				{
				    'fieldname': 'quality_status_yard',
				    'label': ('Quality Status yard'),
				    'fieldtype': 'Data',
				    # 'options': 'Product Order Details'
					'width':'150'
				},
				
    		]
	return col


def get_data(filters=None):
	#{'company': 'DataSoft', 'report_date': '2022-12-12', 'item': 'item finshed', 'product_order': 'PO00001', 'ageing_based_on': 'Due Date', 'range1': 30, 'range2': 60, 'range3': 90, 'range4': 120}
	results = []
	if filters:
		if filters.product_order:
			product_O_details = frappe.get_doc("Product Order", filters.product_order).as_dict()
			for product_detail in product_O_details.product_details:
				# print(product_detail.sap_serial_no)
				obj = {
						'product_order': filters.product_order,
						'item_serial': product_O_details.item_serial,
						'pallet_no':product_detail.pallet_no,
						'quality_status':product_detail.quality_status,
						'item_status':product_detail.item_status,
						'creation_date':product_detail.creation,
						'sap_serial': product_detail.sap_serial_no,
						'qt_inspection_lab':product_detail.qt_inspection_lab,
						'qt_inspection_yard':product_detail.qt_inspection_yard,
						'quality_status_lab':product_detail.quality_status_lab,
						'quality_status_yard':product_detail.quality_status_yard

					}
				results.append(obj)
				
		else:
			product_orders = frappe.get_all("Product Order",fields=['name', 'creation'])

			for product_order in product_orders: 
				if (product_order.creation).strftime('%Y-%m-%d') >= filters.from_date and (product_order.creation).strftime('%Y-%m-%d') <= filters.to_date:
					PO_productDetails = frappe.get_doc("Product Order", product_order.name).as_dict()
					for product_detail in PO_productDetails.product_details:
						# print(product_detail)
						obj = {
							'product_order': product_order.name,
							'item_serial': PO_productDetails.item_serial,
							'pallet_no':product_detail.pallet_no,
							'quality_status':product_detail.quality_status,
							'item_status':product_detail.item_status,
							'creation_date':product_detail.creation,
							'sap_serial': product_detail.sap_serial_no
						}
						results.append(obj)
	else:
		product_orders = frappe.get_all("Product Order")
		for product_order in product_orders:
			PO_productDetails = frappe.get_doc("Product Order", product_order.name).as_dict()
			for product_detail in PO_productDetails:
				obj = {
					'product_order': product_order.name,
					'item_serial': PO_productDetails.item_serial,
					'pallet_no':product_detail.pallet_no,
					'quality_status':product_detail.quality_status,
					'item_status':product_detail.item_status,
					'creation_date':product_detail.creation,
					'sap_serial': product_detail.sap_serial_no
				}

				results.append(obj)
			#{'name': 'dec22b9f76', 'owner': 'Administrator', 'creation': datetime.datetime(2023, 1, 18, 15, 52, 44, 983276), 'modified': datetime.datetime(2023, 2, 14, 10, 55, 10, 731396), 'modified_by': 'ali@datasoft@eg.com', 'docstatus': 0, 'idx': 10, 'ref': 'MRQCBLHS30-11900-230', 'pallet_no': '0', 'net_weight': None, 'gross_weight': None, 'jambo_roll_no': None, 'quality_status': None, 'item_status': 'Waiting Quality', 'qt_inspection': None, 'scaler': None, 'qr_code': None, 'sap_serial_no': '563/10', 'sap_pallet_no': '563/0', 'qt_inspection_lab': None, 'qt_inspection_yard': None, 'quality_status_lab': None, 'quality_status_yard': None, 'qt_final': None, 'parent': 'PO01104', 'parentfield': 'product_details', 'parenttype': 'Product Order', 'doctype': 'Product Order Details'}
	
	
	return results

File: sap/sap/report/product_order_script_report/product_order_script_report.js
// Copyright (c) 2023, ahmed and contributors
// For license information, please see license.txt
/* eslint-disable */
var item_name ="";
function get_item_name(itemCode){
	
	frappe.call({
		method: 'alquds.alqudsQueries.get_Item_name',
		args:{
			item_code : itemCode
		},
		callback(r){
			
			item_name = r.message;
			
		}
	})
}
frappe.query_reports["Product Order Script Report"] = {
	"filters": [
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"reqd": 1,
			"default": frappe.defaults.get_user_default("Company")
		},
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today()
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today()
		},
		{
			"fieldname": "item",
			"label": __("Items"),
			"fieldtype": "Link",
			"options": "Item",
			on_change: () => {
				var itemCode = frappe.query_report.get_filter_value('item');
				get_item_name(itemCode);
				var itemName = item_name;
				return {
					filters: {
						'item_serial': itemName
					}
				}
			}
		},
		// {
		// 	"fieldname": "product_order_details",
		// 	"label": __("Product Order Details"),
		// 	"fieldtype": "Link",
		// 	"options": "Product Order Details",
		// },
		{
			"fieldname": "product_order",
			"label": __("Product Order"),
			"fieldtype": "Link",
			"options": "Product Order",
			get_query: () => {
				return {
					filters: {
						'item_serial': item_name
					}
				}
			}
		},
		// {
		// 	"fieldname": "supplier",
		// 	"label": __("Supplier"),
		// 	"fieldtype": "Link",
		// 	"options": "Supplier",
		// 	on_change: () => {
			
		// 	}
		// },
		// {
		// 	"fieldname": "party_account",
		// 	"label": __("Payable Account"),
		// 	"fieldtype": "Link",
		// 	"options": "Account",
		// 	get_query: () => {
		// 		var company = frappe.query_report.get_filter_value('company');
		// 		return {
		// 			filters: {
		// 				'company': company,
		// 				'account_type': 'Payable',
		// 				'is_group': 0
		// 			}
		// 		};
		// 	}
		// },
		// {
		// 	"fieldname": "ageing_based_on",
		// 	"label": __("Ageing Based On"),
		// 	"fieldtype": "Select",
		// 	"options": 'Posting Date\nDue Date\nSupplier Invoice Date',
		// 	"default": "Due Date"
		// },
		// {
		// 	"fieldname": "range1",
		// 	"label": __("Ageing Range 1"),
		// 	"fieldtype": "Int",
		// 	"default": "30",
		// 	"reqd": 1
		// },
		// {
		// 	"fieldname": "range2",
		// 	"label": __("Ageing Range 2"),
		// 	"fieldtype": "Int",
		// 	"default": "60",
		// 	"reqd": 1
		// },
		// {
		// 	"fieldname": "range3",
		// 	"label": __("Ageing Range 3"),
		// 	"fieldtype": "Int",
		// 	"default": "90",
		// 	"reqd": 1
		// },
		// {
		// 	"fieldname": "range4",
		// 	"label": __("Ageing Range 4"),
		// 	"fieldtype": "Int",
		// 	"default": "120",
		// 	"reqd": 1
		// },
		// {
		// 	"fieldname": "payment_terms_template",
		// 	"label": __("Payment Terms Template"),
		// 	"fieldtype": "Link",
		// 	"options": "Payment Terms Template"
		// },
		// {
		// 	"fieldname": "supplier_group",
		// 	"label": __("Supplier Group"),
		// 	"fieldtype": "Link",
		// 	"options": "Supplier Group"
		// },
		// {
		// 	"fieldname": "group_by_party",
		// 	"label": __("Group By Supplier"),
		// 	"fieldtype": "Check"
		// },
		// {
		// 	"fieldname": "based_on_payment_terms",
		// 	"label": __("Based On Payment Terms"),
		// 	"fieldtype": "Check",
		// },
		// {
		// 	"fieldname": "show_remarks",
		// 	"label": __("Show Remarks"),
		// 	"fieldtype": "Check",
		// },
		// {
		// 	"fieldname": "tax_id",
		// 	"label": __("Tax Id"),
		// 	"fieldtype": "Data",
		// 	"hidden": 1
		// },
		// {
		// 	"fieldname": "show_future_payments",
		// 	"label": __("Show Future Payments"),
		// 	"fieldtype": "Check",
		// }
	]
};
